Ability to cancel a Navigation event

MariovanZeist opened the pull request:
Added a LocationChanging event on the NavigationManager so users are able to cancel navigation events.

LocationChanging events can be canceled for calling NavigationManager.NavigateTo directly (except when setting the forceload flag), 
Clicking an href element
Performing a history.back or history.forward function or clicking the back / forward history buttons in the browser.

For performance reasons and to minimize Server/browser interactions (Especially when using ServerSide Blazor) 
the JS Runtime will only call into C# code when an event listener has been added otherwise every navigation event would incur an extra Client/Server roundtrip when presumably in most cases no Listener will be registered.

Addresses #23886

MariovanZeist committed:
Implement LocationChangingEvent

MariovanZeist committed:
Added E2E Test for cancelation

mrpmorris commented:
Wow :)

MariovanZeist committed:
Merged from master

captainsafia commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -4,34 +4,45 @@ import { EventDelegator } from '../Rendering/EventDelegator';
 
 let hasEnabledNavigationInterception = false;
 let hasRegisteredNavigationEventListeners = false;
+let hasLocationChangingListeners = false;
```
Nit: The formatting in this file needs to be corrected.

MariovanZeist replied:
Can you elaborate? I am not sure what formatting you want changed.

captainsafia replied:
I noticed a few minor things like the braces on L91 to L100, L145, and some additional whitespace on L134.

captainsafia commented on src/Components/test/testassets/BasicTestApp/RouterTest/LocationChangingCancel.razor:
```diff
@@ -0,0 +1,46 @@
+
+<button id="test-button" @onclick=@( () => NavigationManager.NavigateTo("LocationChangingStart"))>Navigate</button>
+
+<a id="test-link" href="LocationChangingStart">Can We</a>>
```
Can we add another test scenario for a link to an external site here (e.g. "google.com")?

MariovanZeist replied:
I will add that test

MariovanZeist replied:
Test have been added

captainsafia reviewed (commented):
Looks good overall! Left some feedback inline.

MariovanZeist committed:
Update src/Components/Components/src/NavigationManager.cs

Co-authored-by: Safia Abdalla <safia@safia.rocks>

MariovanZeist commented on src/Components/test/E2ETest/Tests/RoutingTest.cs:
```diff
@@ -434,6 +434,41 @@ public void UriHelperCanReadAbsoluteUriIncludingHash()
         }
 
+        [Fact]
+        public void CanCancelLocationChanging()
```
I will add a test.

MariovanZeist replied:
Tests have been added

MariovanZeist committed:
Merge remote-tracking branch 'upstream/master' into LocationChangingEvent_MvZ

MariovanZeist committed:
Resynced with master

MariovanZeist committed:
Merge branch 'LocationChangingEvent_MvZ' of https://github.com/MariovanZeist/aspnetcore into LocationChangingEvent_MvZ

MariovanZeist committed:
Source cleanup

MariovanZeist committed:
Added 4 more tests, for canceling external and invalid routes.

MariovanZeist commented:
Api review and recommendations.
--------------------------------------

When the forceLoad flag is true in the function
```cs
public void NavigateTo(string uri, bool forceLoad = false)
```

I have opted to **not** call the OnLocationChanging event, but I am not sure if that was a good decision.

The main use case I foresee for the LocationChanging event will be the _"There are unsaved data on this page, do you want to save"_ scenarios.
In those cases, it might still be better to call the OnLocationChanging event, so the code can take some action.

I can think of 3 ways we could implement this.

1. Leave as is. When the developer chooses to forceLoad, he probably doesn't want the OnLocationChanged to be fired.
2. Call OnLocationChanged, but when it returns ForceLoad anyway. (So ignore the cancel)
3. Call OnLocationChanged but respect Cancel.

I personally lean towards option 3, because that gives the developer the most flexibility, but I can find good arguments for any of these options.

In Cases 2 and 3 we should Add the ForceLoad flag to the OnLocationChanged event, so the developer knows what was intended.
We might want to Combine this with my proposal in https://github.com/dotnet/aspnetcore/pull/25752 by using
NavigationOptions but that is just a thought

ShaunCurtis commented:
Forgive me for asking this (polite) question if I've misunderstood what your trying to do.  Why are you trying to change NavigationManager to modify routing behaviour?  The NavigationManager is just the messenger, exposing JavaScript/HTML navigation events from the loaded page into Blazor.  If no one's registered with LocationChanged, then nothing happens.   In a normal configuration, Router is registered, as part of the App load, and executes view changes and page loads.  By view changes I mean routing and page loads I mean URL based full page loads.  So if you want to make route/not route decisions, make it in the router,

There's further discussion here [issues/14962](https://github.com/dotnet/aspnetcore/issues/14962) and a link to some work I've done on the subject.

ShaunCurtis commented:
Forgive me for asking this (Polite) question if I've misunderstood the above discussion.  Why are you trying to modify NavigationManager?  NavigationManager is just the messenger, surfacing Javascript/HTML navigation events into Blazor.  NavigationManager doesn't router/reload the page, Router does.  So if you are looking to modify/manage router behaviour, do it in the router.

There's another issue here [issues/14962](https://github.com/dotnet/aspnetcore/issues/14962) covering this subject and a linbk to some work I've done on it.

MariovanZeist commented:
@ShaunCurtis This PR Adds a LocationChanging event, which is fired before A LocationChanged event.
It's added to the NavigationManager because that's the earliest opportunity to intercept and cancel the navigation request, without it ever firing a LocationChanged event. The Router reacts to LocationChanged events and that's too late to cancel the navigation without introducing some side effects, As at that point the only way to "cancel" a route change is to Navigate again to the previous page, so instead of not navigating, you end up Navigating twice.

Some of these Side effects include.
1. Multiple calls To LocationChanged, Users that Hook into the LocationChanged event will see these events which may trigger unwanted code. (example would be a call to an Api to retrieve some data)
2. Browser History state is compromised. Every time you cancel route navigation this way, 2 more entries are added to the browser history state. 

So if you want to cancel a routechange because for example your model is dirty you could use something like this. 

```cs
    public interface IPageState
    {
        bool IsDirty { get; }
    }

    public class CheckDirty : ComponentBase, IDisposable
    {
        [Inject] NavigationManager NavigationManager { get; set; }
        [Inject] IPageState PageState { get; set; }
        protected override void OnInitialized()
        {
            NavigationManager.LocationChanging += HandleLocationChanging;
        }

        private void HandleLocationChanging(object? sender, LocationChangingEventArgs e)
        {
            if (PageState.IsDirty)
            {
                e.Cancel = IsDirty;
            }
        }

        public void Dispose()
        {
            NavigationManager.LocationChanging -= HandleLocationChanging;
        }
    }
```

mrpmorris commented:
@MariovanZeist I prefer `args.Cancel ()` so another handler can't assume it's the only interested party and set it back to false.

MariovanZeist commented:
@mrpmorris I can see the benefits of your proposal. I am going to add that to my Api Review section above, so as to see what other people think about it. The main reason I had for the Cancel being a settable property was consistency with other EventArgs. But that may be an inadequate justification.

captainsafia commented:
Thanks for prepping this list of API review items, @MariovanZeist! We'll be doing an API review for this on Monday and will report back.

ShaunCurtis commented:
Hi Mario,

Thank you for taking the time to reply.  I had forgotten about the two trip issue to fix the displayed url - I shouldn't have as I had to write code to deal with it!.  I agree with your logic.   When you write code for the current framework you sometimes don't raise your perspective high enough to see the whole picture.  What got me started on this subject was the whole issue of dirty model data, and what is and is not dirty a dirty model - there are other issues with the edit form controls. 

On ForceLoad, I would respect cancel. I can see scenarios in teams where the main navigation controls, such as a left side nav bar are maintained by a junior programmer who sets ForceLoad because he doesn't know any better!  If I have a dirty model in an edit form I don't want to exit without user confirmation. period.

SteveSandersonMS commented:
Thanks for this, @MariovanZeist! This is really impressive. Thanks for doing the E2E tests as well!

--- 

Regarding this part: https://github.com/dotnet/aspnetcore/pull/24417/files#diff-b761f1cc117bcbb43cdf980fda82ee8dR157

That seems a bit problematic, both because it's discarding "forward" entries, and because it's pushing new things onto the history stack that mean the back/forward history will contain unexpected entries if navigation is later allowed.

There is a different way this could be approached. Instead of using `history.pushState` to revert, you could use `history.go(delta)`. This allows the forward/back stack to be retained completely. Here's a sketch of how it could work: https://gist.github.com/SteveSandersonMS/1924c7c0c46a6434518665924b2a97b4. Do you think we should consider that kind of approach?

---

Besides that, about the `forceLoad` parameter. I see there's open discussion about whether force-loaded navigations should be cancellable or not. Rather than having to guess what developers want, would it work to include the "forceLoad" flag on the OnNavigating event args, so that developers can choose how to account for it in their own "block-or-not" logic?

MariovanZeist committed:
merged from master

MariovanZeist committed:
Fix for uninitialized JSRuntime
When we arrive from an External Page into a Page which has a NavigationManager and has a LocationChanging Event registered,
Server Side Blazor would crash because of the uninitialized JSRuntime, during prerendering

MariovanZeist committed:
Fix for Early initialzation problem.
When we add LocationChanging eventhandlers before the JSRuntime was initialized, the RemoteNavigationManager would crash.
Also the navigation subsystem would not be informed we had Eventhandlers unless that state changed.

MariovanZeist commented:
@pranavkm Sorry I didn't see the Working label until I pushed some changes. 
I have some other issues I am working on (implementing @SteveSandersonMS suggestions)
You want me to hold off on working on those changes for now?

pranavkm commented:
Sorry, I accidentally marked a PR as working.

MariovanZeist committed:
LocationChanging event is now also called when ForceLoad is set in NavigateTo

MariovanZeist committed:
Added E2E test for Forceload=True

MariovanZeist committed:
Merge remote-tracking branch 'upstream/master' into LocationChangingEvent_MvZ

MariovanZeist committed:
Updated publicAPI

MariovanZeist commented:
@captainsafia I am trying to create an E2E test to validate the fix I made at 82605a13fe869bbcf99ffea7995319cfdcfb8544 

The test would be similar to the ```CanGoBackFromNotAComponent```, But it needs an initialized Navigationmanager, according to this remark in the above-mentioned tests:
https://github.com/dotnet/aspnetcore/blob/fe6a3fc747096308ac3acdf7d72437f6955490d5/src/Components/test/E2ETest/Tests/RoutingTest.cs#L350-L353

I am not sure if this is possible. Can you advise?

MariovanZeist committed:
Merge remote-tracking branch 'upstream/master' into LocationChangingEvent_MvZ

captainsafia commented on src/Components/Components/src/NavigationManager.cs:
```diff
@@ -30,6 +30,29 @@ public abstract class NavigationManager
+        /// <summary>
+        /// An event that fires when the navigation location is about to change
+        /// </summary>
+        public event EventHandler<LocationChangingEventArgs> LocationChanging
```
Feedback from API review: we discussed the risks of implementing the `LocationChanging` API as an event handler, especially with regard to async scenarios. For example, if a user wants to send an HTTP request to determine whether the users content has been saved before navigating.

Instead of event, we discussed maintaining an internal list of of navigation listeners.

```csharp
private List<Func<LocationChangingContext, ValueTask>> _locationChangingListeners ;
```

The `LocationChangingContext` property above contains the same properties as the `LocationChangingEventArgs` but is renamed since we're no longer implementing the event handler pattern here. The listener the user provides will return a `ValueTask` that can be awaited.

Since the `_locationChangingListeners` property is internal, we discussed implementing a `RegisterLocationChangingListener` method on the `NavigationManager` that implements the following API.

```csharp
public void RegisterLocationChangingListener(Func<LocationChangingContext, ValueTask> listener)
```

When invoked, this method will add the provided `listener` to the internally maintained list.

When it comes to using the registered ` _locationChangingListeners`, we'll await each of the listeners in the handle location changing API. If any one of the listeners returns `false`, we do not await the remaining listeners.

MariovanZeist replied:
Good catch, even marking an event handler as async and awaiting a call would not have had the intended effect, due to Blazor's SynchronizationContext.

I do have some additional remarks about the above approach. I preface my remarks with my expectation that the main use cases for the LocationChanging 'event' is the ability of a user to either cancel the current navigation or do some emergency housekeeping just before navigation would occur.

I would want to keep the ability to add and remove listeners so that users can add listeners on a per component base level. The above proposal hints towards a global registration of handlers. (I base my observation on the fact that you are using a register function, and ASPNET Core doesn't use the Register/Unregister pair, and that you are registering a Func<>)

My second remark is about:
> If any one of the listeners returns false, we do not await the remaining listeners.

~~I think this might introduce some confusion and would invalidate some use cases (In particular the use case where a user would do some emergency housekeeping (on LocationChanging).
I would either always call the registered listeners, or add an additional boolean to indicate that this listener should always be called even when another listener canceled.~~

_After some more personal deliberation, I would agree with your proposal._ 


My changed proposal would be something like.

```cs
public class LocationChangingContext 
{
  public string Location { get; }
  public bool IsNavigationIntercepted { get; }
}
```

Adding an Interface
```cs
public interface ILocationChanging
{
  ValueTask<bool> OnLocationChanging(LocationChangingContext context);
}
```

Add some functions to ```NavigationManager```
```cs
  public void AddLocationChangingListener(ILocationChanging locationChanging);
  public bool RemoveLocationChangingListener(ILocationChanging locationChanging);
```

I prefer an interface instead of
```cs
  Func<LocationChangingContext, ValueTask<bool>>
```
Because it can be easily removed. and users cannot simply use something like
```cs
  Add( async (ctx) =>await ShouldWeCancel() );
```
which would prevent users from removing the above function, and could introduce memory leaks.

MariovanZeist replied:
I am going to be a little bold, and mark this as resolved, this does not by any means mean that the discussion is over, but I have committed code that implements this API change roughly as I described above. There are some questions I have, but I will add that to another comment, And feel free to criticize my implementation.

captainsafia reviewed (commented):
Thanks for your patience with this API review, @MariovanZeist!

We took a look at this and posted some ideas on how to improve this. Let me know if you have any follow-up questions or need help with any of the follow-on work.

MariovanZeist committed:
Changed LocationChanging implementation from an Event to a Handler system, so we can use async/await

MariovanZeist commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -4,34 +4,45 @@ import { EventDelegator } from '../Rendering/EventDelegator';
+let hasLocationChangingListeners = false;
+
+// Current uri
+let currentUri ='';
```
I am working on some other changes related to this, especially steves comment: https://github.com/dotnet/aspnetcore/pull/24417#issuecomment-706236021 There are some corner cases I need to consider to properly implement this, so it isn't a straightforward change. So this part is still a Work In Progress

MariovanZeist commented on src/Components/Server/src/Circuits/RemoteNavigationManager.cs:
```diff
@@ -64,9 +65,19 @@ public void NotifyLocationChanged(string uri, bool intercepted)
+
         /// <inheritdoc />
-        protected override void NavigateToCore(string uri, bool forceLoad)
+        protected override async void NavigateToCore(string uri, bool forceLoad)
```
As we want async/await capabilities when handling a LocationChange I had to modify this function to an ```async void``` (Yikes 🤔)  so too not to introduce breaking changes to the current API Surface. 
The best solution would be to not only change this function to an ```async Task``` , but to also Change ```void Navigationmanager.NavigateTo``` to ```Task Navigationmanager.NavigateTo```. 
So I would like some advice on how we would want to proceed. (The ```WebAssemblyNavigationManager``` has a similar approach )

captainsafia commented:
@MariovanZeist and I chatted a bit about the potential confusion that we would have in the framework if we were to merge this PR. We'd have two different concepts around navigation/location changing: the `OnNavigateAsync` attribute we added in the router as part of the lazy-loading work and the `LocationChanging` listener in the `NavigationManager`. I worry that having these two similar-yet-different concepts in the framework will be confusing to people.

We should figure out a way to reconcile the two concepts before merging this PR. @MariovanZeist Maybe we should head back to the original issue for this and hash out a design that accounts for:

- The distinction between navigation cancellation and a general location changing event
- What to do with the `OnNavigateAsync` method now that we're introducing this API

ChristianWeyer commented:
Do we have any indication of whether this will make it into 6.0 or not? .cc @mkArtakMSFT

javiercn commented:
@ChristianWeyer It's in our radar and I believe we plan to get it in, however currently the team is working on other bigger items (Hot reload, AoT, Blazor Desktop, etc.)

ChristianWeyer commented:
OK, thanks for the quick update. Much appreciated.

captainsafia commented:
@ChristianWeyer Yep -- since this feature does have a fair bit of complexity we're taking some time to draft up a design doc (https://github.com/dotnet/aspnetcore/issues/28527) so that we've covered all bases before implementation.
